src/ray/raylet/node_manager_test.cc

```diff
@@ -0,0 +1,159 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <thread>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "ray/common/test_util.h"
+#include "ray/raylet/node_manager.h"
+
+namespace ray {
+
+namespace raylet {
+
+class TestNodeManager : public ::testing::Test {
+ public:
+  TestNodeManager() {
+    // Start redis.
+    TestSetupUtil::StartUpRedisServers(std::vector<int>());
+
+    // Start store.
+    store_socket_name_ = TestSetupUtil::StartObjectStore();
+
+    // Create object manager config.
+    object_manager_config_.store_socket_name = store_socket_name_;
+    object_manager_config_.pull_timeout_ms = 1;
+    object_manager_config_.object_chunk_size = static_cast<uint64_t>(std::pow(10, 3));
+    object_manager_config_.push_timeout_ms = 1000;
+    object_manager_config_.object_manager_port = 0;
+    object_manager_config_.rpc_service_threads_number = 3;
+
+    // Create node manager config.
+    node_manager_config_.heartbeat_period_ms = 100000;
+    node_manager_config_.store_socket_name = store_socket_name_;
+    node_manager_config_.maximum_startup_concurrency = 1;
+    node_manager_config_.node_manager_address = "127.0.0.1";
+    node_manager_config_.min_worker_port = 1000;
+    node_manager_config_.max_worker_port = 30000;
+  }
+
+  virtual ~TestNodeManager() {
+    // Stop store.
+    TestSetupUtil::StopObjectStore(store_socket_name_);
+
+    // Stop redis.
+    TestSetupUtil::ShutDownRedisServers();
+  }
+
+  void SetUp() {
+    server_io_service_.reset(new boost::asio::io_service());
+    server_io_service_thread_.reset(new std::thread([this] {
+      std::unique_ptr<boost::asio::io_service::work> work(
+          new boost::asio::io_service::work(*server_io_service_));
+      server_io_service_->run();
+    }));
+
+    client_io_service_.reset(new boost::asio::io_service());
+    client_io_service_thread_.reset(new std::thread([this] {
+      std::unique_ptr<boost::asio::io_service::work> work(
+          new boost::asio::io_service::work(*client_io_service_));
+      client_io_service_->run();
+    }));
+
+    // Start redis gcs client.
+    gcs::GcsClientOptions client_options("127.0.0.1", TEST_REDIS_SERVER_PORTS.front(),
+                                         /*password*/ "",
+                                         /*is_test_client=*/true);
+    gcs_client_ = std::make_shared<gcs::RedisGcsClient>(client_options);
+    RAY_CHECK_OK(gcs_client_->Connect(*client_io_service_));
+
+    // Create object manager.
+    NodeID node_id = NodeID::FromRandom();
+    object_manager_.reset(new ObjectManager(
```

Comment:
Why don't we mock out this class? Is it because code is not written in a way that it is unit-testable?

Reply:
I tried mocking it at first, but its constructor couldn't mock it. So most of UT's logic is for its constructor to execute successfully. Later, if the test needs, we can mock it.

Reply:
I think we don't need to test anything else other than the placement group logic, so it should be fine! (since the new scheduler will be UT'ed and object manager also will be UT'ed in the future. The only thing we need to test is the idempotency of bundle requests.)